typing/typedecl_variance.ml

```diff
@@ -263,60 +291,82 @@ let compute_variance_extension env ~check decl ext rloc =
     {decl with type_params = ext.ext_type_params}
     (ext.ext_args, ext.ext_ret_type)
 
-let compute_variance_decl env ~check decl (required, _ as rloc) =
+let compute_variance_decl env ~check id decl (required, _ as rloc) =
+  let check =
+    if check
+    then Some (Type_declaration (id,decl))
+    else None
+  in
   if (decl.type_kind = Type_abstract || decl.type_kind = Type_open)
        && decl.type_manifest = None then
     List.map
       (fun (c, n, i) ->
         make (not n) (not c) (decl.type_kind <> Type_abstract || i))
       required
-  else
-  let mn =
-    match decl.type_manifest with
-      None -> []
-    | Some ty -> [false, ty]
-  in
-  let vari = match decl.type_kind with
-    Type_abstract | Type_open ->
-      compute_variance_type env ~check rloc decl mn
-  | Type_variant (tll,_rep) ->
-      if List.for_all (fun c -> c.Types.cd_res = None) tll then
-        compute_variance_type env ~check rloc decl
-          (mn @ List.flatten (List.map (fun c -> for_constr c.Types.cd_args)
-                                tll))
-      else begin
-        let mn =
-          List.map (fun (_,ty) -> (Types.Cstr_tuple [ty],None)) mn in
-        let tll =
-          mn @ List.map (fun c -> c.Types.cd_args, c.Types.cd_res) tll in
-        match List.map (compute_variance_gadt env ~check rloc decl) tll with
-        | vari :: rem ->
-            List.fold_left (List.map2 Variance.union) vari rem
-        | _ -> assert false
-      end
-  | Type_record (ftl, _) ->
-      compute_variance_type env ~check rloc decl
-        (mn @ List.map (fun {Types.ld_mutable; ld_type} ->
-             (ld_mutable = Mutable, ld_type)) ftl)
-  in
-  if mn = [] || decl.type_kind <> Type_abstract then
-    List.map Variance.strengthen vari
-  else vari
+  else begin
+    let mn =
+      match decl.type_manifest with
+        None -> []
+      | Some ty -> [ false, ty ]
+    in
+    let vari =
+      match decl.type_kind with
+        Type_abstract | Type_open ->
+          compute_variance_type env ~check rloc decl mn
+      | Type_variant (tll,_rep) ->
+          if List.for_all (fun c -> c.Types.cd_res = None) tll then
+            compute_variance_type env ~check rloc decl
+              (mn @ List.flatten (List.map (fun c -> for_constr c.Types.cd_args)
+                                    tll))
+          else begin
+            let vari =
+              match decl.type_manifest with
+              | None -> []
+              | Some ty ->
+                  let mn = Types.Cstr_tuple [ ty ], None in
+                  [ compute_variance_gadt env ~check rloc decl mn ]
+            in
```

Comment:
Now that the code cannot structured to pretend that the `type_manifest` is another constructor due to the improved error message, I am not sure if it makes sense to call `compute_variance_gadt`. Inlining the various definitions gives us
```ocaml
let vari =
  List.map (compute_variance_type env ~check rloc { decl with type_private = Private}) mn
in
```